lib/geckodriver.js

```diff
@@ -0,0 +1,152 @@
+    } = args;
+
+    this.proxyHost = host;
+    this.proxyPort = port;
+    this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort, base: ''});
+    this.state = Geckodriver.STATE_STOPPED;
+    this.capabilities = {};
+  }
+
+  async start (caps, emitStartingState = true) {
+    this.capabilities = _.cloneDeep(caps);
+
+    const args = [`--host=${this.proxyHost}`, `--port=${this.proxyPort}`, '-vv'];
+
+    const startDetector = (stdout) => {
+      return stdout.indexOf('Listening') !== -1;
+    };
+
+    if (emitStartingState) {
+      this.changeState(Geckodriver.STATE_STARTING);
+    }
+
+    let processIsAlive = false;
+    try {
+      this.geckodriver = SYSTEM_GECKODRIVER;
+      this.proc = new SubProcess(this.geckodriver, args);
+      processIsAlive = true;
+
+      this.proc.on('output', (stdout, stderr) => {
+        if (this.verbose) {
+          for (let line of (stdout || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.debug(`[STDOUT] ${line}`);
+          }
+          for (let line of (stderr || '').trim().split('\n')) {
+            if (!line.trim().length) continue; // eslint-disable-line curly
+            log.error(`[STDERR] ${line}`);
+          }
+        }
+      });
+
+      this.proc.on('exit', (code, signal) => {
+        log.info(code, signal);
+        processIsAlive = false;
+
+        if (this.state !== Geckodriver.STATE_STOPPED &&
+          this.state !== Geckodriver.STATE_STOPPING &&
+          this.state !== Geckodriver.STATE_RESTARTING) {
+          let msg = `Geckodriver exited unexpectedly with code ${code}, ` +
+                  `signal ${signal}`;
+          log.error(msg);
+          this.changeState(Geckodriver.STATE_STOPPED);
+        }
+      });
+
+      log.info(`Spawning geckodriver with: ${this.geckodriver} ` +
+              `${args.join(' ')}. Only geckodriver 0.26.0 supports Android testing`);
+      // start subproc and wait for startDetector
+      await this.proc.start(startDetector);
+      this.changeState(Geckodriver.STATE_ONLINE);
+      await this.startSession();
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+      if (processIsAlive) {
+        await this.proc.stop();
+      }
+    }
+  }
+
+  sessionId () {
+    if (this.state !== Geckodriver.STATE_ONLINE) {
+      return null;
+    }
+    return this.jwproxy.sessionId;
+  }
+
+  async stop () {
+    try {
+      await this.proc.stop();
+      this.changeState(Geckodriver.STATE_STOPPED);
+    } catch (e) {
+      this.emit(Geckodriver.EVENT_ERROR, e);
+    }
+  }
+
+  async getStatus () {
+    return await this.jwproxy.command('/status', 'GET');
+  }
+
+  prepareMozCapabilities () {
+    let mozFirefoxOptions = {
+      log: {
+        level: 'trace'
+      },
+      androidPackage: 'org.mozilla.firefox',
+      androidActivity: 'org.mozilla.gecko.BrowserApp',
+      androidDeviceSerial: 'FA6980300148',
```

Comment:
this should not be here, right?

Reply:
@mykola-mokhnach I'm removing the androidDeviceSerial, but the package and activity are needed

Reply:
device serial should also be present in caps for the case multiple devices are connected to the same server